[Amber Tan Junning] iP

ambertan77 opened the pull request:
### Introducing BOB!!! 🥳

Way too many tasks to keep track of? Bob helps you remember everything! Bob is:
- text-based
- really easy to use
- **SUPER** FAST
- **_FREE_ TO USE!!!!**

Here's what you need to do to get started:
1. Download the jar file [here](https://github.com/ambertan77/ip/releases/tag/v0.1)!
2. Double-click it
3. Add your tasks
4. And Bob will **remember & manage** every single task added for you 🤩

Features:
- [X] Adding and removing tasks
- [X] Marking and unmarking tasks
- [X] Finding tasks using search keys
- [X] Listing all you tasks 
- [ ] Managing deadlines (coming soon!)
- [ ] Reminders (coming soon!)

Lastly, Bob wants to say...
> "Remember, never despise any task - any task. In doing any job, you are either creating something or bettering something" - Janette Oke [(source)](https://www.goodreads.com/quotes/10071114-remember-never-despise-a-task-any-task-in-doing-any-job)

If you are a Java programmer, you can use it to practice `Java` too. Here's the main method:
```
public static void main(String[] args) throws Exception {
    new Bob("./data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ambertan77 committed:
implement greeting and exit messages

ambertan77 committed:
implement echoing and exit command

ambertan77 committed:
implement storing and displaying user input

ambertan77 committed:
add task class

ambertan77 committed:
add mark and unmark features

ambertan77 committed:
add deadline class

ambertan77 committed:
add event class

ambertan77 committed:
add todos class

ambertan77 committed:
make minor edits to deadline class

ambertan77 committed:
fix minor typos in todos class

ambertan77 committed:
add support for tracking todos, deadlines and events

ambertan77 committed:
fix minor typos in greeting

ambertan77 committed:
fix minor typos in string output for addition of new tasks

ambertan77 committed:
add string header for list

ambertan77 committed:
change name of main class

ambertan77 committed:
add test cases for supported user commands

ambertan77 committed:
add exceptions to deal with incorrect input

ambertan77 committed:
change data type that stores the list of tasks

ambertan77 committed:
implement deleting command

ambertan77 committed:
add code comments and clean up code

ambertan77 committed:
add enum for types of tasks

ambertan77 committed:
add enum for supported user commands

ambertan77 committed:
Format code to meet coding standard

ambertan77 committed:
Add getters to the Task class and its subclasses

ambertan77 committed:
Bob.java: Add logic to create, write and append to tasks file

Previously, there was no logic to save the task list into a file in the
hard disk automatically.

Such logic will allow users to remember what they have previously input.

Let's,
* add the relevant logic to create the new directory and file if needed
* add a method to write to the file
* add a method to append text to the file

ambertan77 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Bob.java: Add logic to create, write and append to tasks file
  Add getters to the Task class and its subclasses

ambertan77 committed:
Bob.java: Add references for code

ambertan77 committed:
Bob.java: Load data from hard disk when chatbot is loaded up

Previously, there was no logic to support loading data from saved on the
hard disk.

Such logic will allow users to be able to access the list of tasks sent
to the chatbot previously.

Let's,
* add a method to add file contents into the current task list
* add a helper method to create tasks from the string inputs scanned
* add file contents to task list whenever data.txt file already exists

ambertan77 committed:
Fix typo errors that occur when appending text to file in hard disk

ambertan77 committed:
Change type of time-sensitive fields from String to LocalDateTime

ambertan77 committed:
Bob.java: Add support for reading user-input date/time strings

ambertan77 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Bob.java: Add support for reading user-input date/time strings
  Change type of time-sensitive fields from String to LocalDateTime
  Fix typo errors that occur when appending text to file in hard disk
  Bob.java: Load data from hard disk when chatbot is loaded up

ambertan77 committed:
Add TaskList class

ambertan77 committed:
Add support for TaskList class

ambertan77 committed:
Add Storage class

ambertan77 committed:
Add support for Storage class

ambertan77 committed:
Add Parser class

ambertan77 committed:
Add support for Parser class

ambertan77 committed:
Add method for loading file when chatbot first starts

ambertan77 committed:
Fix minor typos when printing date/time

ambertan77 committed:
Add Ui class

ambertan77 committed:
Add support for Ui class

ambertan77 committed:
Organise classes into packages

ambertan77 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

ambertan77 committed:
Fix typo in code comments

ambertan77 committed:
Remove constructor for Ui class

ambertan77 committed:
Parser.java: Create different methods for different types of tasks

ambertan77 committed:
Fix the return type for methods creating the different types of tasks

ambertan77 committed:
Add instructions on the format to follow for creation of Deadline tasks

ambertan77 committed:
Build.gradle: Change file names

ambertan77 committed:
Add tests for the createDeadlineTask method in Parser.java

ambertan77 committed:
Storage.java: Handle tasks incorrectly formatted in file

ambertan77 committed:
Storage.java: Edit minor typos

ambertan77 committed:
Add tests for the createTaskFromFile method in Storage.java

ambertan77 committed:
Specify name of jar file

ambertan77 committed:
Deadline.java: Add JavaDoc comments

ambertan77 committed:
Deadline.java: Fix typos in JavaDoc comments

ambertan77 committed:
Event.java: Add JavaDoc comments

ambertan77 committed:
Add JavaDoc comment for constructors

ambertan77 committed:
Todos.java: Add JavaDoc comments

ambertan77 committed:
Task.java: Add JavaDoc comments

ambertan77 committed:
Bob.java: Add JavaDoc comments

ambertan77 committed:
Parser.java: Add JavaDoc comments

ambertan77 committed:
Storage.java: Add JavaDoc comments

ambertan77 committed:
TaskList.java: Add JavaDoc comments

ambertan77 committed:
Ui.java: Add JavaDoc comments

ambertan77 committed:
Change import order to match coding standards

ambertan77 committed:
Task.java: Place conditionals on a separate line

ambertan77 committed:
TaskList.java: Add logic to find tasks by keyword

ambertan77 committed:
Add support for the find command

ambertan77 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Ui.java: Add JavaDoc comments
  TaskList.java: Add JavaDoc comments
  Storage.java: Add JavaDoc comments
  Parser.java: Add JavaDoc comments
  Bob.java: Add JavaDoc comments
  Task.java: Add JavaDoc comments
  Todos.java: Add JavaDoc comments
  Add JavaDoc comment for constructors
  Event.java: Add JavaDoc comments
  Deadline.java: Fix typos in JavaDoc comments
  Deadline.java: Add JavaDoc comments

ambertan77 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Task.java: Place conditionals on a separate line
  Change import order to match coding standards

# Conflicts:
#	src/main/java/bob/Bob.java
#	src/main/java/bob/Parser.java
#	src/main/java/bob/Storage.java
#	src/main/java/bob/TaskList.java

ambertan77 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add support for the find command
  TaskList.java: Add logic to find tasks by keyword

ambertan77 committed:
TaskList.java: Add JavaDoc comments for the find method

ambertan77 committed:
Add files to check style of code

ambertan77 committed:
build.gradle: Add checkstyle plugins

ambertan77 committed:
Parser.java: Add whitespace in line 126

ambertan77 committed:
Parser.java: Add default case at line 293

ambertan77 committed:
Change import order for test files

ambertan77 committed:
StorageTest.java: Add whitespace in try-catch blocks

ambertan77 committed:
Storage.java: Remove additional whitespace in line 109

ambertan77 committed:
Storage.java: Change order of variables decalared

ambertan77 committed:
Todos.java: Remove unused import statement

ambertan77 committed:
Add files required for checkstyle

ambertan77 committed:
Merge branch 'branch-A-Checkstyle'

* branch-A-Checkstyle:
  Add files required for checkstyle
  Todos.java: Remove unused import statement
  Storage.java: Change order of variables decalared
  Storage.java: Remove additional whitespace in line 109
  StorageTest.java: Add whitespace in try-catch blocks
  Change import order for test files
  Parser.java: Add default case at line 293
  Parser.java: Add whitespace in line 126
  build.gradle: Add checkstyle plugins

MakiseKurisuX commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,298 @@
+            Task task = createTask(input);
+            tasks.add(task);
+            try {
+                String filePath = "./data/tasks.txt";
```
Any reason why in this file you opted to hardcode the file location instead of using the filePath from Bob.java?

MakiseKurisuX commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,298 @@
+            Task task = createTask(input);
+            tasks.add(task);
+            try {
+                String filePath = "./data/tasks.txt";
```
Any reason as to why you hardcoded filePath in this file instead of using the filePath from Bob.java?

ambertan77 replied:
No reason actually. I think it will be better to follow the filePath from Bob.java to ensure consistency in case the file containing the data has changed. Thank you for the suggestion!

MakiseKurisuX commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,183 @@
+     * Adds the tasks stored in the data file into the current task list.
+     *
+     * @throws Exception If the file cannot be read or if the data is formatted wrongly.
+     */
```
I like how you added comments for all public classes and methods, and added references where useful.

MakiseKurisuX commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,298 @@
+                System.out.println("Unable to write to file: " + e.getMessage());
+            }
+            System.out.println(unmark + "   " + taskToUnmark.toString() + "\n" + line);
+            return;
```
Consider extracting the marking and unmarking action as separate methods within TaskList as you may need to add exception handling for invalid numbers.

```suggestion
        case MARK:
            // mark task as done
            System.out.println(line);
            int indexToMark = Integer.valueOf(input.substring(5));
            tasks.mark(indexToMark - 1) // suggested change
            try {
                String filePath = "./data/tasks.txt";
                Task firstTask = tasks.get(0);
                storage.writeToFile(filePath, firstTask);
                for (int i = 1; i < tasks.count; i++) {
                    Task task = tasks.get(i);
                    storage.appendToFile(filePath, task);
                }
            } catch (IOException e) {
                System.out.println("Unable to write to file: " + e.getMessage());
            }
            System.out.println(mark + "   " + taskToMark.toString() + "\n" + line);
            return;
        case UNMARK:
            // mark task as not done
            System.out.println(line);
            int indexToUnmark = Integer.valueOf(input.substring(7));
            tasks.unmark(indexToUnmark - 1) // suggested change
            try {
                String filePath = "./data/tasks.txt";
                Task firstTask = tasks.get(0);
                storage.writeToFile(filePath, firstTask);
                for (int i = 1; i < tasks.count; i++) {
                    Task task = tasks.get(i);
                    storage.appendToFile(filePath, task);
                }
            } catch (IOException e) {
                System.out.println("Unable to write to file: " + e.getMessage());
            }
            System.out.println(unmark + "   " + taskToUnmark.toString() + "\n" + line);
            return;
```

MakiseKurisuX commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,298 @@
+                }
+            } catch (IOException e) {
+                System.out.println("Unable to write to file: " + e.getMessage());
+            }
```
Perhaps you could encapsulate this as a method? It is used several times in your commands.

ambertan77 replied:
Sounds good, thank you!

MakiseKurisuX commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,183 @@
+public class Storage {
+
+    protected boolean isNewFile = false;
+    private TaskList tasks;
```
Well-named variable as it clearly indicates that it stores a list of tasks.

MakiseKurisuX commented on src/main/java/bob/Ui.java:
```diff
@@ -0,0 +1,58 @@
+        scanner.close();
+        System.out.println(exit);
+    }
+}
```
I like the way you did the UI extraction in accordance to OOP principles.

MakiseKurisuX commented on src/main/java/bob/task/Event.java:
```diff
@@ -0,0 +1,83 @@
+        return "[E]" + super.toString() + " (from " + this.getFrom() + " to " + this.getTo() + ")";
+    }
+
+    /**
+     * Returns the string representation of the type of task.
+     *
+     * @return Letter representation of the type of task.
+     */
+    public String getType() {
+        return "E";
+    }
```
Since this extends from Task, perhaps you may want to add an Override? This comment applies to your other objects where there is no Override notation.

```suggestion
    /**
     * Returns the string representation of the type of task.
     *
     * @return Letter representation of the type of task.
     */
    @Override
    public String getType() {
        return "E";
    }
```

ambertan77 replied:
Noted, thank you for the suggestion.

ambertan77 committed:
Update build.gradle to set up JavaFX

ambertan77 committed:
Create custom exception for Bob

ambertan77 committed:
TaskList.java: Create a new method to get the current count

ambertan77 committed:
BobException.java: Add reference for header comments

ambertan77 committed:
Parser.java: Throw BobException instead of Exception

ambertan77 committed:
Task.java: Throw BobException instead of Exception

ambertan77 committed:
Ui.java: Throw BobException instead of Exception

ambertan77 committed:
Bob.java: Throw BobException instead of Exception

ambertan77 committed:
Allow users to command "list" before tasks are added to the list

ambertan77 committed:
Storage.java: Throw BobException instead of Exception

ambertan77 committed:
build.gradle: Change main class

ambertan77 committed:
Add images used in GUI

ambertan77 committed:
Change return type of methods used to interact with user

ambertan77 committed:
Ui.java: Add method to greet user

ambertan77 committed:
Create GUI for Bob

ambertan77 committed:
Change user image

ambertan77 committed:
Refactor code to use FXML and add related classes to a new package

Previously, the logic behind the GUI for Bob was all within the Main
class and DialogBox. Using FXML will help make the code more concise.

Additionally, putting all the classes related to the GUI in the
same package helps to improve the organization of code.

ambertan77 committed:
build.gradle: Change file path of main class

ambertan77 committed:
Create command: fix minor typo in output

ambertan77 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Create command: fix minor typo in output
  build.gradle: Change file path of main class
  Refactor code to use FXML and add related classes to a new package
  Change user image
  Create GUI for Bob
  Ui.java: Add method to greet user
  Change return type of methods used to interact with user
  Add images used in GUI
  build.gradle: Change main class
  Storage.java: Throw BobException instead of Exception
  Allow users to command "list" before tasks are added to the list
  Bob.java: Throw BobException instead of Exception
  Ui.java: Throw BobException instead of Exception
  Task.java: Throw BobException instead of Exception
  Parser.java: Throw BobException instead of Exception
  BobException.java: Add reference for header comments
  TaskList.java: Create a new method to get the current count
  Create custom exception for Bob
  Update build.gradle to set up JavaFX

kaungzinye commented on src/main/java/bob/task/Event.java:
```diff
@@ -0,0 +1,83 @@
+     * @return End time of event.
+     */
+    public String getTo() {
+        DateTimeFormatter outputStringFormat = DateTimeFormatter.ofPattern("dd/MM/yyyy, HH:mm");
```
Consider making the function to obtain the output string for the dates to use a singular DateTimeFormatter to reduce repeating code as well as to make sure that the output format of both are the same, reducing the risk of possible bugs

ambertan77 replied:
That's a great idea, thank you for the suggestion!

kaungzinye reviewed (commented):
LGTM. Just some small details that would allow for more consistency

ambertan77 committed:
Add assertions to code to if assumptions hold

Several methods have implicit assumptions of specific properties in
parameters and variables.

It will be better to check if these assumptions hold, before moving on
to run the other parts of the code. This helps the debugging process
since the code will stop running at the point where one assumption made
does not hold true.

ambertan77 committed:
Add Override annotations for methods in the children classes

ambertan77 committed:
Use one DateTimeFormatter in the class

To ensure consistency in the DateTimeFormatter used in different methods
in the class, add the DateTimeFormatter as a global variable so that
different methods can access it.

ambertan77 committed:
Add method to retrieve the file path used

File path has been hardcoded in different classes in the code. To ensure
consistency, add a method to get the file path by this instance of Bob.
This also reduces the room for errors to be made during hardcoding.

ambertan77 committed:
Add getters and setters to protected fields

ambertan77 committed:
Refactor code to represent commands as classes and package them together

ambertan77 committed:
Edit test class to align with the refactored code

ambertan77 committed:
Extract logic to improve readability

The createTaskFromFile method handles different cases, causing it to be
long and messy.

Extracting the logic implemented in the different cases
improves code readability.

Let's create helper methods and use them in the
createTaskFromFile method.

ambertan77 committed:
Use streams for find method

ambertan77 committed:
Merge pull request #1 from ambertan77/branch-A-Assertions

Add assertions to code to check if assumptions hold

ambertan77 committed:
Merge branch 'master' into branch-A-CodeQuality

ambertan77 committed:
Merge pull request #2 from ambertan77/branch-A-CodeQuality

Improve code quality

ambertan77 committed:
Merge pull request #3 from ambertan77/branch-A-Streams

Use Java Streams in TaskList

ambertan77 committed:
Add gradle.yml to set up Continuous Integration

ambertan77 committed:
Remove I/O redirection tests

ambertan77 committed:
Edit code to conform to coding conventions

ambertan77 committed:
Edit caught exceptions in test methods

ambertan77 committed:
Fix minor typo in text saved onto hard disk

ambertan77 committed:
Add the ability to check and remove duplicates

There is no restriction on what tasks can be added into the task list,
other than the input format specified.

This allows duplicate tasks, with the exact same details and
descriptions, to be added.

Let's
* add a command that allows users to check for duplicates
* add a command that allows users to remove all duplicates

ambertan77 committed:
Add support for the checking and removing of duplicate tasks

ambertan77 committed:
Add header comments for new methods created

ambertan77 committed:
Merge branch 'branch-BCD-Extension'

* branch-BCD-Extension:
  Add header comments for new methods created
  Add support for the checking and removing of duplicate tasks
  Add the ability to check and remove duplicates
  Fix minor typo in text saved onto hard disk

ambertan77 committed:
Handle all exceptions thrown

Some exceptions thrown are not handled, which causes the program to
become unusable.

When exceptions are caught in the UI and error messages are returned as
text messages from Bob, users will have a better understanding of the
error that has occurred.

Let's update the methods in Bob.java and Ui.java to handle all
exceptions by returning the error messages as Strings.

ambertan77 committed:
Edit error messages to provide additional details

ambertan77 committed:
Handle IndexOutOfBoundsException during deletion of tasks

ambertan77 committed:
Add title bar for the GUI

ambertan77 committed:
Edit code to keep to line length specified in coding standards

ambertan77 committed:
Edit code to comply with coding standards

ambertan77 committed:
Update expected error messages in test cases

ambertan77 committed:
Edit header message for the delete command

ambertan77 committed:
Change dimensions of GUI display

davidfoo07 commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+    protected int count = 0;
+
+    // store the list of tasks
+    protected ArrayList<Task> tasks;
```
Good declaration of multi-valued variables.

davidfoo07 commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+
+    // store the list of tasks
+    protected ArrayList<Task> tasks;
+    protected HashMap<Task, ArrayList<Task>> duplicates = new HashMap<>();
```
Also, good declaration of multi-valued variables

davidfoo07 commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return A string containing the greeting message, or error message if an error has occurred.
+     */
+    public String run() {
```
Method follows the verb clause.

davidfoo07 commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,78 @@
+     *
+     * @return The file path for the file stored in the hard disk.
+     */
+    public String getFilePath() {
```
Clear naming of method regarding its function.

davidfoo07 commented on src/main/java/bob/commands/CreateCommand.java:
```diff
@@ -0,0 +1,208 @@
+    public static final String CREATE_HEADER = "Got it. I've added this task:\n";
+
+    private TaskList tasks;
+    private Storage storage;
```
Variable follows the noun clause.

davidfoo07 commented on src/main/java/bob/TaskList.java:
```diff
@@ -0,0 +1,133 @@
+ */
+public class TaskList {
+    // store the current count of tasks in the list
+    protected int count = 0;
```
Slight improvement would be renaming count to taskCount to be more specific on its reference to the number of task lists.

davidfoo07 commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,265 @@
+     * @return The newly created task.
+     * @throws BobException If the file cannot be read or if the data is formatted wrongly.
+     */
+    public Task createTaskFromFile(String storedInput) throws BobException {
```
These create SOMETHING from file methods name can be refined to parse SOMETHING from file to focus on the idea that you are reading or parsing from a file

ambertan77 committed:
README.md: Add user guide

ambertan77 committed:
Use clearer product screenshot

ambertan77 committed:
Resize product screenshot image in README.md

ambertan77 committed:
Edit section links in README.md

ambertan77 committed:
Fix minor typo in README.md

ambertan77 committed:
Handle invalid date and time inputs

Invalid dates and times that have been input are not handled by the
CreateCommand class.

This can cause the program to stop working as an uncaught exception will
be thrown.

Let's handle the exception thrown when invalid dates and times have been
input.

ambertan77 committed:
Add test cases for inputs containing invalid dates and times

ambertan77 committed:
Edit Quick Start section in README.md

ambertan77 committed:
Add code reference for test cases

ambertan77 committed:
Edit minor typos in README.md

ambertan77 committed:
Edit code to comply with coding standards

ambertan77 committed:
Set Markdown processor to GFM

ambertan77 committed:
Add additional code references for test cases

ambertan77 committed:
Fix bug in delete command

ambertan77 committed:
Storage.java: Fix minor typos in error description

ambertan77 committed:
Fix bug in writing to file in hard disk

ambertan77 committed:
Edit User Guide to fix minor typos
